DoublyLinkedList.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * DoublyLinkedList.c
+ *
+ *  Created on: Nov 3, 2022
+ *      Author: Mohamed Refat
+ */
+
+#include <stdlib.h>
+#include "STD_TYPES.h"
+#include "DoublyLinkedList.h"
+#include <assert.h>
+void DoublyLinkedList_Init(DoublyLinkedList_t *pList)
+{
+	pList->pHead = NULL_ptr;
+	pList->pTail = NULL_ptr;
+	pList->Size  = 0;
+}
+
+uint8_t DoublyLinkedList_InsertNode(DoublyLinkedList_t *pList , uint32_t Position , LINKEDLIST_TYPE Data )
+{
+	Node_t *pLocNode ;
+	Node_t *pLocTempNodeNode ;
+	uint8_t LocReturn;
+
+
+	if( (pLocNode = (Node_t *)malloc(sizeof(Node_t))) )
+	{
+		pLocNode->Data = Data ;
+		if(pList->pTail ==NULL_ptr && pList->pHead == NULL_ptr)
+		{
+			pList->pTail = pLocNode ;
+			pList->pHead = pLocNode ;
+		}
+
+		if ( Position==0 )
+		{
+			pLocNode->pNext = pList->pHead ;
+			pList->pHead->pPrevious = pLocNode ;
+			pLocNode->pPrevious = NULL_ptr ;
+			pList->pHead = pLocNode ;
+
+		}else if ( Position == (pList->Size-1) )
+		{
+			pLocNode->pPrevious = pList->pTail ;
+			pLocNode->pNext     = NULL_ptr ;
+			pList->pTail->pNext = pLocNode ;
+			pList->pTail = pLocNode ;
+
+		}else if(Position < pList->Size)
+		{
+			pLocTempNodeNode = pList->pHead ;
+			uint16_t i ;
+			for(i=0 ; i<Position-1 ; i++)
+			{
+				pLocTempNodeNode = pLocTempNodeNode->pNext ;
+			}
+			pLocNode->pPrevious = pLocTempNodeNode ;
+			pLocNode->pNext     = pLocTempNodeNode->pNext ;
+			pLocTempNodeNode->pNext->pPrevious = pLocNode ;
+			pLocTempNodeNode->pNext = pLocNode ;
+
+		}else if(Position >= pList->Size)
+		{
+			assert( (!(Position >= pList->Size))  && "The position larger then the size , so enter the right position ");
+		}else
+		{
+			LocReturn = FALSE ;
+		}
+		pList->Size++;
+		LocReturn =  TRUE ;
+	}else
+	{
+		LocReturn =  FALSE ;
+	}
+
+	return LocReturn ;
+}
+uint8_t  DoublyLinkedList_InsertAfter(DoublyLinkedList_t *pList , LINKEDLIST_TYPE Data)
+{
+	Node_t *pLocNode;
+	uint8_t LocReturn;
+
+	if( (pLocNode = (Node_t *)malloc(sizeof(Node_t))) )
```

Comment:
Your implementation is totally right but it can be tricky for another developer or for you after long time as you used to see logical operator == inside if condition so by default you (or another developer) will change it and your code will fail😢
To git rid of this complex situation you have to make sth like that :

```
  pLocNode = (Node_t *)malloc(sizeof(Node_t);
  if( pLocNode == 1) {
```